autoload/intero/process/backend_info.vim

```diff
@@ -0,0 +1,49 @@
+"""""""""""
+" Version:
+"
+" This file contains functions for parsing and checking backend information,
+" e.g. the GHCi or Intero version.
+"""""""""""
+
+" A backend info contains the backend identifier, which is 'ghci' or 'intero',
+" and a version. A version is a list of three numbers, for the MAJOR, MINOR, and
+" PATCH versions.
+function! intero#process#backend_info#backend_info(backend, major, minor, patch) abort
+    let l:version = [a:major + 0, a:minor + 0, a:patch + 0]
+    return {'backend': a:backend, 'version': l:version}
+endfunction
+
+" Tries to parse the backend info from a list of lines. Returns a dict with
+" backend info if successful, otherwise an empty dict.
+function! intero#process#backend_info#parse_lines(output) abort
+    for l:l in a:output
+        " Try parsing regular GHCi version.
+        let l:matches = matchlist(l:l, 'GHC\%(i\|\sInteractive\), version \(\d*\)\.\(\d*\).\(\d*\)[:,]')
+        if !empty(l:matches)
+            return intero#process#backend_info#backend_info(
+                        \ 'ghci',
+                        \ l:matches[1],
+                        \ l:matches[2],
+                        \ l:matches[3])
+        else
+            " Fallback to parsing Intero-style version.
+            let l:matches = matchlist(l:l, 'Intero .* (GHC \(\d*\)\.\(\d*\).\(\d*\))')
```

Comment:
Nit: if we defined a list of regexes and iterated over them, this could would be much shorter.

(We could also combine the regexes, but that probably wouldn't be great for readability)